### PR TITLE
Create an initial password for new users

### LINKED ISSUE #76: Create an initial password when creating a user
This is less a feature request than a mergeability poll, as I'm very much inclined to submit a pull request for this.

My feature would be

+ When a user is created and `initial_password` is set
A call to `/users/{user_id}reset-password` is made to initialize a password.

+ When a user is updated and `initial_password` is set
No further call is made

+ When `initial_password` is not set, do as is done now.

Would you accept such a feature ?

### PR DESCRIPTION
**Pull request by Floby**
Fixes #76

I got to the point where I can login with a newly created user.

However, I don't know how to deal with the diffing from an existing user which is why my test continues to fail. Reading through the [terraform docs](https://godoc.org/github.com/hashicorp/terraform/helper/schema#Resource) for the Resource didn't help me much. I would like some help with your terraform-fu @mrparkers :]

Anyway, I'd like a review on my code as it is my first time writing actual go.

### CODE REVIEW

File: provider/keycloak_user_test.go [hunk 1]
```diff
@@ -34,6 +39,36 @@ func TestAccKeycloakUser_basic(t *testing.T) {
...
+					testAccCheckKeycloakUserExists(resourceName),
+					testAccCheckKeycloakUserInitialPasswordLogin(realmName, username, password, clientId),
+				),
+			},
+			{
+				ResourceName:        resourceName,
+				ImportState:         true,
+				ImportStateVerify:   true,
+				ImportStateIdPrefix: realmName + "/",
+			},
```
[the remark this answers is not in the document]
Comment (Floby) [reply provider/keycloak_user_test.go]: Great ! It's clearer to me now what this actually tests. I'll remove it.

File: provider/keycloak_user_test.go [hunk 2]
```diff
@@ -227,6 +262,38 @@ func testAccCheckKeycloakUserFetch(resourceName string, user *keycloak.User) res
...
+		if err != nil {
+			return err
+		}
+		request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+		response, err := httpClient.Do(request)
+		if err != nil {
+			return err
+		}
+
```
[the remark this answers is not in the document]
Comment (Floby) [reply provider/keycloak_user_test.go]: I've seen this in code samples but couldn't figure what it did so I left it out. Guess this takes care of asynchronicity and so on.

File: provider/keycloak_user_test.go [hunk 3]
```diff
@@ -227,6 +262,38 @@ func testAccCheckKeycloakUserFetch(resourceName string, user *keycloak.User) res
 	}
 }
 
+func testAccCheckKeycloakUserInitialPasswordLogin(realmName string, username string, password string, clientId string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		httpClient := &http.Client{}
+
+		resourceUrl := os.Getenv("KEYCLOAK_URL") + "/auth/realms/" + realmName + "/protocol/openid-connect/token"
```
[the remark this answers is not in the document]
Comment (Floby) [reply provider/keycloak_user_test.go]: sure thing

File: provider/keycloak_user_test.go [hunk 4]
```diff
@@ -281,6 +348,34 @@ resource "keycloak_user" "user" {
...
+    client_id           = "%s"
+
+    name                = "test client"
+    enabled             = true
+
+    access_type         = "PUBLIC"
+    direct_access_grants_enabled = true
+    valid_redirect_uris = [
+        "*"
+    ]
```
[the remark this answers is not in the document]
Comment (Floby) [reply provider/keycloak_user_test.go]: haha, alright.

File: provider/keycloak_user.go [hunk 5]
```diff
@@ -81,6 +86,14 @@ func resourceKeycloakUserCreate(data *schema.ResourceData, meta interface{}) err
 		return err
 	}
 
+	_, isPasswordSet := data.GetOk("initial_password")
```
[the remark this answers is not in the document]
Comment (Floby) [reply provider/keycloak_user.go]: yeah, thought about that afterwards on the train :D

**Comment by Floby** [comment]
I tried quickly with `DiffSuppressFunc` but couldn't figure out how to use it. I have a hard time debugging and figuring out when the function is called and with what arguments :/

**Comment by mrparkers** [comment]
The function signature looks like this:

```go
type SchemaDiffSuppressFunc func(resourceName, old, new string, data *ResourceData) bool
```

The goal here would be to suppress all diffs for this attribute after the resource has been created.  You can use `data` to see if the resource has an ID - if it does, you know the resource has been created, because Keycloak is responsible for generating those IDs.

**Comment by Floby** [comment]
Aaah great, I missed that fact. I was only trying to check with IsNew.

Thanks

Le jeu. 3 janv. 2019 à 18:34, Michael Parker <notifications@github.com> a
écrit :

**Comment by Floby** [comment]
Hello @mrparkers,
I was able to supress the diff when updating a existing user. I also added (and documented) a flag `initial_password_temporary` which can set up the password to be renewed on first login. Could not figure how to add a test simply though.

I'm happy with how the feature works now but I'm still happy to get feedback.

**Comment by mrparkers** [comment]
Thanks for addressing the comments.  The code looks good, although I am not sure how I feel about the new `initial_password_temporary` field.  It seems weird to have a companion attribute that is only relevant if another attribute is also set.

If you want to keep that functionality, what would you think of this instead?

```hcl
resource "keycloak_user" "user_with_password" {
	realm_id   = "${keycloak_realm.test.id}"
	username   = "user-with-password"

	email      = "user-with-password@fakedomain.com"
	first_name = "Testy"
	last_name  = "Tester"

	initial_password {
            value     = "my-password"
            temporary = true # defaults to false
        }
}
```

I have to admit, I am not that familiar with implementing this type of schema, but I do think it looks cleaner.  Would you be okay with taking a stab at this?

If not, I'd prefer that we just make a decision to hardcode the initial password to either always be temporary or not (whatever is better for your use case), and I can open an issue and try to do this at a later time.

Let me know what you think.

**Comment by Floby** [comment]
I agree it would be cleaner. As a matter of fact I tried several things but I could not find many examples around.

You can either :
+ Declare `initial_password` to be a `schema.TypeMap` but the keys are then arbitrary which would make for less usefull (absent in fact) error messages when using it wrongly
+ Declare `initial_password` to be a `schema.TypeSet` which allows to have a `schema.Resource` as the `Elem` property. But this allows to have an arbitrary number of declarations for `initial_password` which seems a bit silly. I guess I could go this way and throw a custom error if this is declared more than once.

**Comment by Floby** [comment]
It seems this is the way to go https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_mq_broker.go#L42-L61

**Comment by mrparkers** [comment]
Yes, it seems everyone uses `TypeList` for stuff like this, which I agree is strange.  I often reference the Google provider and it does this as well.

**Comment by Floby** [comment]
Got it to work with a block declaration !

File: docs/resources/keycloak_user.md [hunk 6]
```diff
@@ -31,6 +46,11 @@ The following arguments are supported:
 
 - `realm_id` - (Required) The realm this user belongs to.
 - `username` - (Required) The unique username of this user.
+- `initial_password` (Optional) When given, the user's initial password will be set
+   This attribute is only respected during initial user creation.
+  - `value` (Required) The initial password
+  - `temporary` (Optional) If set to `true`, the initial password is set up for renewal on first use. Default to `false`.
+- `initial_password_temporary` (Optional) Forces the initial password to be renewed on first login. Default to `false`.
```
Comment (mrparkers) [inline docs/resources/keycloak_user.md]: This shouldn't be needed anymore

File: docs/resources/keycloak_user.md [hunk 7]
```diff
@@ -31,6 +46,11 @@ The following arguments are supported:
 
 - `realm_id` - (Required) The realm this user belongs to.
 - `username` - (Required) The unique username of this user.
+- `initial_password` (Optional) When given, the user's initial password will be set
+   This attribute is only respected during initial user creation.
```
Comment (mrparkers) [inline docs/resources/keycloak_user.md]:
```suggestion
   This block attribute is only respected during initial user creation. This block can be configured with the following attributes:
```

File: docs/resources/keycloak_user.md [hunk 8]
```diff
@@ -31,6 +46,11 @@ The following arguments are supported:
 
 - `realm_id` - (Required) The realm this user belongs to.
 - `username` - (Required) The unique username of this user.
+- `initial_password` (Optional) When given, the user's initial password will be set
```
Comment (mrparkers) [inline docs/resources/keycloak_user.md]:
```suggestion
- `initial_password` (Optional) When given, the user's initial password will be set.
```

File: docs/resources/keycloak_user.md [hunk 9]
```diff
@@ -31,6 +46,11 @@ The following arguments are supported:
...
+   This attribute is only respected during initial user creation.
+  - `value` (Required) The initial password
```
Comment (mrparkers) [inline docs/resources/keycloak_user.md]:
```suggestion
  - `value` (Required) The initial password.
```

File: provider/keycloak_user_test.go [hunk 10]
```diff
@@ -281,6 +371,34 @@ resource "keycloak_user" "user" {
...
+	direct_access_grants_enabled = true
+}
+
+resource "keycloak_user" "user" {
+	realm_id         = "${keycloak_realm.realm.id}"
+	username         = "%s"
+  initial_password {
+    value = "%s"
+    temporary = false
+  }
```
Comment (mrparkers) [inline provider/keycloak_user_test.go]: nitpick: could you use tabs here?

**Comment by Floby** [comment]
Hello, I think we're close to good here :)
Do you need these commits squashed ?

**Comment by Floby** [comment]
Thank you for bearing with me and accepting my work :) Your repo is neat and very tidy. Very easy to get into.

Do you plan on making a new release soon ?

**Comment by mrparkers** [comment]
I appreciate the kind words!  I'm about to cut a new release for you right now.  It'll be ready in 10 minutes or so.

**Comment by Floby** [comment]
Great ! thanks a lot :)

**Pull request merged**